### ISSUE TITLE
plot learning curve

### ISSUE THREAD

**Issue by xuzhang5788** [dreamquark-ai/tabnet#261]
## Feature request
history of learning. plot loss vs. epochs for train data and validation data

**What is the expected behavior?**

**What is motivation or use case for adding/changing the behavior?**
visualize the learning process

**How should this be implemented in your opinion?**

**Are you willing to work on this yourself?**
yes

**Comment by Optimox** [comment]
All the metrics at each epoch are available via the ‘model.history’, you should be able to plot most of the things you want.

**Comment by Optimox** [comment]
@xuzhang5788 was the history what you were looking for? Can I close this? Or is it something different?

**Comment by xuzhang5788** [comment]
It will be great if you would like to provide an example. Thanks

**Comment by SeohuiPark** [comment]
Hi, first of all, thank you for the great code! I'm going to use this model for my project, and I have a few questions. 
I know that validation performance can be checked through eval_metric. 

But how can I calculate the **validation loss?** 

When I wrote the pytorch code, I set it as "model.eval()" and added the process of calculating loss, but does the eval_metric of Tabnet not provide it?

**Comment by Optimox** [comment]
@SeohuiPark you can create a custom metric which is the training loss like cross entropy or MSE.

**Comment by micheusch** [comment]
so, when modifying https://github.com/dreamquark-ai/tabnet/blob/develop/forest_example.ipynb to add this 

```
class cross_entropy_reporting(Metric):
    def __init__(self):
        self._name = "c_e" # write an understandable name here
        self._maximize = False
    def __call__(self, y_true, y_score):
        return float(torch.nn.functional.cross_entropy(torch.tensor(y_score), torch.tensor(y_true)))
```

and using it in training

```
clf.fit(
    X_train=X_train, y_train=y_train,
    eval_set=[(X_train, y_train), (X_valid, y_valid)],
    eval_name=['train', 'valid'],
    eval_metric=['c_e', 'accuracy'],
    loss_fn=torch.nn.functional.cross_entropy,
    max_epochs=max_epochs, patience=100,
    batch_size=16384, virtual_batch_size=256,
    augmentations=aug
) 
```
I get quite different loss function values. 
```
epoch 0  | loss: 1.25592 | train_c_e: 1.92593 | train_accuracy: 0.23095 | valid_c_e: 1.92764 | valid_accuracy: 0.22855 |  0:00:11s
epoch 1  | loss: 0.77832 | train_c_e: 1.92819 | train_accuracy: 0.23431 | valid_c_e: 1.92521 | valid_accuracy: 0.23725 |  0:00:23s
epoch 2  | loss: 0.71297 | train_c_e: 1.96275 | train_accuracy: 0.20043 | valid_c_e: 1.96155 | valid_accuracy: 0.20149 |  0:00:35s
epoch 3  | loss: 0.69437 | train_c_e: 1.99151 | train_accuracy: 0.16943 | valid_c_e: 1.98909 | valid_accuracy: 0.17192 |  0:00:47s
epoch 4  | loss: 0.6839  | train_c_e: 2.04683 | train_accuracy: 0.10996 | valid_c_e: 2.04694 | valid_accuracy: 0.1097  |  0:00:59s
epoch 5  | loss: 0.67324 | train_c_e: 1.99889 | train_accuracy: 0.12821 | valid_c_e: 1.99854 | valid_accuracy: 0.12906 |  0:01:11s
epoch 6  | loss: 0.6619  | train_c_e: 1.97077 | train_accuracy: 0.14489 | valid_c_e: 1.9706  | valid_accuracy: 0.14521 |  0:01:23s
epoch 7  | loss: 0.65169 | train_c_e: 1.94991 | train_accuracy: 0.15938 | valid_c_e: 1.94961 | valid_accuracy: 0.16016 |  0:01:35s
```

Even considering slight difference between average on batches and evaluation at epoch end, difference looks a bit big between loss and train_c_e. Am I missing something silly?

**Comment by Optimox** [comment]
I think the error comes from the activation function: the loss function is applied to the outputs before activation, cross entropy = LogSoftmax + NLLoss. However the outputs that you get for metric computation are Softmax of logits. So you should try something like :

```
class cross_entropy_reporting(Metric):
    def __init__(self):
        self._name = "c_e" # write an understandable name here
        self._maximize = False
    def __call__(self, y_true, y_score):
        return float(torch.nn.functional.NLLLoss(torch.log(torch.tensor(y_score)), torch.log(torch.tensor(y_true))))
```

Also note that they are two other reasons why the train loss and train_c_e won't exactly match:
- average of batches loss != overall loss
- the loss you see also takes into account the sparsity loss, so if you want to have results as similar as possible you should set `lambda_sparse=0`

**Comment by micheusch** [comment]
oh nice one, I had figured last two, but missed NLLLoss, thanks for clarifying!
